File: iOS/Kalorie/Core/UseCases/DeleteFoodItemReportUseCase.swift
//
//  DeleteFoodItemReportUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 17.09.2026.
//

import Foundation

protocol DeleteFoodItemReportUseCaseProtocol {
    func callAsFunction(barcode: String, reportedBy: String) async throws
}

struct DeleteFoodItemReportUseCase: DeleteFoodItemReportUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(barcode: String, reportedBy: String) async throws {
        guard authProvider.userId != nil else { throw AuthError.notAuthenticated }
        try await dataProvider.deleteAsync(
            id: FoodItemReportDomain.id(barcode: barcode, userId: reportedBy),
            from: Constants.Firestore.foodItemReports
        )
    }
}

#if DEBUG
struct DeleteFoodItemReportUseCaseFake: DeleteFoodItemReportUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(barcode: String, reportedBy: String) async throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/DeleteFoodItemReportUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemReportDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.utils.Constants

interface DeleteFoodItemReportUseCaseProtocol {
    suspend operator fun invoke(barcode: String, reportedBy: String)
}

class DeleteFoodItemReportUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : DeleteFoodItemReportUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(barcode: String, reportedBy: String) {
        if (authProvider.userId == null) throw AuthError.NotAuthenticated
        dataProvider.deleteAsync(
            id = FoodItemReportDomain.id(barcode = barcode, userId = reportedBy),
            from = Constants.Firestore.FOOD_ITEM_REPORTS,
        )
    }
}
